### ISSUE TITLE
Data variables for permalinks broken in *.11ty.js templates

### ISSUE THREAD

**Issue by reubenlillie** [11ty/eleventy#1076]
**Describe the bug**

When using `*.11ty.js` as a content template (i.e., within the _input directory_, not a as a layout in either the includes or layout directories), data variables in permalinks are output as plain text (e.g., `_site/{{page.fileSlug}}/index.html`).

**To Reproduce**

1. Within the input directory, create a directory data file (e.g., `./pages/pages.11tydata.js`) and set the property `permalink: '/{{page.fileSlug}}/index.html'`
2. Create a `*.11ty.js` content template in that same directory (e.g., `./pages/test.11ty.js`)
3. Run `npx eleventy --serve`
4. Observe the following output:

```cli
Writing _site/{{page.fileSlug}}/index.html from ./pages/test.11ty.js.
```

**Expected behavior**

Data variables in permalinks should not conflict with `*.11ty.js` content templates.

Steps 1–3 should produce:

```cli
Writing _site/test/index.html from ./pages/test.11ty.js.
```

**Screenshots**

**Sample directory data file `./pages/pages.11tydata.js`**

![image](https://user-images.githubusercontent.com/16792087/78741306-9b515580-791e-11ea-8f97-0ae65d7ab460.png)

**Sample content template `./pages/test.11ty.js`**

![image](https://user-images.githubusercontent.com/16792087/78741343-b45a0680-791e-11ea-81d6-f3563b7339f7.png)

**Sample layout `_includes/layouts/base.11ty.js` (for reference)**

![image](https://user-images.githubusercontent.com/16792087/78741241-7bba2d00-791e-11ea-88ed-721c17d974bc.png)

**Sample output using `npx eleventy --serve`**

![image](https://user-images.githubusercontent.com/16792087/78741132-33027400-791e-11ea-8236-290ab3bd5f86.png)

The issue does _not_ occur when omitting data variables in permalinks

**Alternate directory data file `./pages/pages.11tydata.js` (without data variables in `permalink`)**

![image](https://user-images.githubusercontent.com/16792087/78743820-6a285380-7925-11ea-9dbf-df0f6df5a497.png)

**Expected output (except, of course, for the preferred permalink structure)**

![image](https://user-images.githubusercontent.com/16792087/78743812-5d0b6480-7925-11ea-8a1f-507d3196cc90.png)

The issue also does _not_ occur when using directory data permalink rules, but calling `*.11ty.js` layouts with other content templates instead (e.g., with Markdown)

**`./pages/alternate.md`, inheriting permalink rules from `./pages/pages.11tydata.js`**

![image](https://user-images.githubusercontent.com/16792087/78744767-0bb0a480-7928-11ea-9401-79ca545fa2f0.png)

**The predictably puzzling output**

![image](https://user-images.githubusercontent.com/16792087/78744802-23882880-7928-11ea-902a-c5cb0c07146e.png)

**Environment:**

- OS and Version: Ubuntu 18.04.4
- Eleventy Version 0.10.0

**Additional context**

I’m optimistic that the forthcoming `eleventyComputed` data feature will allow for JS template literals within `permalink` values, instead of requiring Liquid/Nunjucks syntax. But I’m not sure if that functionality alone will resolve this specific issue.

While hardly ideal, it is _possible_ to achieve the desired output with `*.11ty.js` content templates by setting `permalink: 'test/index.html'` explicitly in the `*.11ty.js` (optional) `data` object. 

**What hackery is this?**

![image](https://user-images.githubusercontent.com/16792087/78742960-eff6cf80-7922-11ea-9f96-e37dcccc2a05.png)

![image](https://user-images.githubusercontent.com/16792087/78742917-cfc71080-7922-11ea-9797-5ab70de58060.png)

**Comment by reubenlillie** [comment]
Related to #642 and/or #943?

**Comment by oliverjam** [comment]
I _think_ this is because permalinks are templated using whatever engine corresponds to the file type.

> These will be parsed with the current template’s rendering engine  
[Use Data Variables in Permalink](https://www.11ty.dev/docs/permalinks/#use-data-variables-in-permalink)

A JS file doesn't have a templating engine, so a string is just a string (no templating happens). If you want a dynamic permalink in a JS file I think you need to use a function:

```js
permalink: data => `/${data.page.fileSlug}/index.html`
```

However there is a related problem (#838) because if you use a non-JS template with that JS layout then Eleventy tries to use whatever templating engine the original _template_ uses (e.g. Liquid for a Markdown file) even though the permalink is defined in a JS file.

**Comment by reubenlillie** [comment]
Sorry I’m so overdue on closing this, @zachleat.

This issue is fixed in v0.11.0 with `eleventyComputed` like so:

```js
// test.11ty.js

exports.data = {
  title: 'Test',
  eleventyComputed: {
    permalink: data => `/${data.page.fileSlug}/index.html`
  }
}
```
